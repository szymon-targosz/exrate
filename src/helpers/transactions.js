import numeral from 'numeral';
import 'numeral/locales';

const getHighest = transactions =>
   transactions.length === 0 ? undefined : transactions.reduce((prev, curr) => (prev.amount > curr.amount ? prev : curr), []);

const getTotal = transactions => transactions.map(transaction => transaction.amount).reduce((sum, val) => sum + val, 0);

const selectTransactions = (transactions, { text, sort }) => {
   return transactions
      .filter(transaction => {
         const textMatch = transaction.name.toLowerCase().includes(text.toLowerCase());
         return textMatch;
      })
      .sort((a, b) => {
         if (sort === 'alpha') {
            return a.name > b.name ? 1 : -1;
         } else if (sort === 'asc') {
            return a.amount - b.amount;
         } else {
            return b.amount - a.amount;
         }
      });
};

const generateAmountMsg = (amount, locale = 'de', decsFit = false) => {
   numeral.locale(locale);

   let decNums = '00';
   if (decsFit) {
      const decs = amount.toString().split('.')[1] || '';
      let i = 0;
      while (i < decs.length - 2) {
         decNums += 0;
         i++;
      }
   }

   return numeral(amount).format(`0,0.${decNums} $`);
};

export { getHighest, getTotal, selectTransactions, generateAmountMsg };
