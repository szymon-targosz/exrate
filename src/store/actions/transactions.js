import uuidv4 from 'uuid/v4';
import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from './actionTypes';

const addTransaction = (name, amount) => ({
   type: ADD_TRANSACTION,
   transaction: {
      id: uuidv4(),
      amount: parseFloat(amount),
      name
   }
});

const removeTransaction = id => ({
   type: REMOVE_TRANSACTION,
   id
});

const removeAllTransactions = () => ({ type: REMOVE_ALL_TRANSACTIONS });

export { addTransaction, removeTransaction, removeAllTransactions };
