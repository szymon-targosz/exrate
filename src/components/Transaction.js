import React from 'react';
import { generateAmountMsg } from '../helpers/transactions';

const Transaction = ({ transaction, rate, removeHandler }) => {
   const { name, amount } = transaction;

   return (
      <div className="transaction">
         <div className="transaction__data">
            <h3 className="heading-3 mb-s">{name}</h3>
            <p className="transaction__value">
               {generateAmountMsg(amount)} / {generateAmountMsg(amount * rate, 'pl')}
            </p>
         </div>
         {removeHandler && (
            <button className="btn btn--link transaction__btn" onClick={removeHandler}>
               Remove
            </button>
         )}
      </div>
   );
};

export default Transaction;
