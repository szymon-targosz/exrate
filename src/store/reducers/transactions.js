import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from '../actions/actionTypes';

export default (state = [], action) => {
   switch (action.type) {
      case ADD_TRANSACTION:
         return state.concat(action.transaction);

      case REMOVE_TRANSACTION:
         return state.filter(({ id }) => id !== action.id);

      case REMOVE_ALL_TRANSACTIONS:
         return [];

      default:
         return state;
   }
};
