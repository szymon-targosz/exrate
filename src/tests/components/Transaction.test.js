import React from 'react';
import Transaction from '../../components/Transaction';
import { shallow } from 'enzyme';
import transactions from '../fixtures/transactions';

let wrapper, removeHandler;

beforeEach(() => {
   removeHandler = jest.fn();
   wrapper = shallow(<Transaction transaction={transactions[0]} rate={4.46} />);
});

describe('COMPONENT - Transaction', () => {
   test('Should render Transaction correctly', () => {
      wrapper.setProps({ removeHandler: removeHandler });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should render Transaction without the remove button', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('Should handle removeTransaction', () => {
      wrapper.setProps({ removeHandler: removeHandler });
      wrapper.find('button').simulate('click');
      expect(removeHandler).toHaveBeenCalled();
   });
});
