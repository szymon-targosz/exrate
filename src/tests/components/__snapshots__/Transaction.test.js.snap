// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - Transaction Should render Transaction correctly 1`] = `
<div
  className="transaction"
>
  <div
    className="transaction__data"
  >
    <h3
      className="heading-3 mb-s"
    >
      Bills
    </h3>
    <p
      className="transaction__value"
    >
      41,00 €
       / 
      182,86 PLN
    </p>
  </div>
  <button
    className="btn btn--link transaction__btn"
    onClick={[MockFunction]}
  >
    Remove
  </button>
</div>
`;

exports[`COMPONENT - Transaction Should render Transaction without the remove button 1`] = `
<div
  className="transaction"
>
  <div
    className="transaction__data"
  >
    <h3
      className="heading-3 mb-s"
    >
      Bills
    </h3>
    <p
      className="transaction__value"
    >
      41,00 €
       / 
      182,86 PLN
    </p>
  </div>
</div>
`;
